cap/error.go

```diff
@@ -3,169 +3,159 @@ package cap
-	return se.nodeErr
+// KVs returns a metadata map for structured logging
+func (err *SupervisorTerminationError) KVs() map[string]interface{} {
+	nodeNames := make([]string, 0, len(err.nodeErrMap))
+	for nodeName := range err.nodeErrMap {
+		nodeNames = append(nodeNames, nodeName)
+	}
+	sort.Strings(nodeNames)
+
+	acc := make(map[string]interface{})
+	acc["supervisor.name"] = err.supRuntimeName
+
+	for i, nodeName := range nodeNames {
+		nodeErr := err.nodeErrMap[nodeName]
+		var subTreeError ErrKVs
+		if errors.As(nodeErr, &subTreeError) {
+			for k0, v := range subTreeError.KVs() {
+				k := strings.TrimPrefix(k0, "supervisor.")
+				acc[fmt.Sprintf("supervisor.subtree.%d.%s", i, k)] = v
+			}
+		} else {
+			acc[fmt.Sprintf("supervisor.termination.node.%d.name", i)] = nodeName
+			acc[fmt.Sprintf("supervisor.termination.node.%d.error", i)] = nodeErr
+		}
+
+	}
+
+	if err.rscCleanupErr != nil {
+		acc["supervisor.termination.cleanup.error"] = err.rscCleanupErr
+	}
+
+	return acc
 }
 
-// GetRuntimeName returns the name of the supervisor that failed
-func (se *SupervisorError) GetRuntimeName() string {
-	return se.supRuntimeName
+// SupervisorBuildError wraps errors returned from a client provided function
+// that builds the supervisor nodes, enhancing it with supervisor information
+type SupervisorBuildError struct {
+	supRuntimeName string
+	buildNodesErr  error
 }
 
-// NodeFailCount returns the number of nodes that failed to terminate correctly.
-// Note if a goroutine fails to terminate because of a shutdown timeout, the
-// failed goroutines may leak. This happens because go doesn't offer any true
-// way to kill a goroutine.
-func (se *SupervisorError) NodeFailCount() int {
-	return len(se.nodeErrMap)
+func (err *SupervisorBuildError) Error() string {
+	return "supervisor build nodes function failed"
 }
 
-// KVs returns a data bag map that may be used in structured logging
-func (se *SupervisorError) KVs() map[string]interface{} {
-	kvs := make(map[string]interface{})
-	kvs["supervisor.name"] = se.supRuntimeName
-	for chKey, chErr := range se.nodeErrMap {
-		kvs[fmt.Sprintf("supervisor.node.%v.stop.error", chKey)] = chErr.Error()
-	}
-	if se.nodeErr != nil {
-		kvs["supervisor.termination.error"] = se.nodeErr.Error()
-	}
-	if se.rscCleanupErr != nil {
-		kvs["supervisor.cleanup.error"] = se.rscCleanupErr.Error()
-	}
-	return kvs
+// KVs returns a metadata map for structured logging
+func (err *SupervisorBuildError) KVs() map[string]interface{} {
+	acc := make(map[string]interface{})
+	acc["supervisor.name"] = err.supRuntimeName
+	acc["supervisor.build.error"] = err.buildNodesErr
+	return acc
+}
+
+// SupervisorStartError wraps an error reported on the initialization of a child
+// node, enhancing it with supervisor information and possible termination errors
+// on other siblings
+type SupervisorStartError struct {
+	supRuntimeName string
+	nodeName       string
+	nodeErr        error
+	terminationErr *SupervisorTerminationError
```

Comment:
nit: Could this be called `childTerminationErr`? I had to read the implementation to understand what this value was not for the current node.

Reply:
Apologies for being a bit pedantic, the terminationErr is abount the current supervisor :grimacing:

A supervisor will have a termination error if any of the children has an error on termination, and a start error may have a termination. I'll add some comments in that line of code and also on the `SupervisorTerminationError` type

Reply:
I was referring to this code: https://github.com/capatazlib/go-capataz/blob/cb2ee3cbd9bee95a92931f4ffe3eff98efd78025/cap/monitor.go#L175-L191

My understanding of the terminology is probably just a bit off.

Reply:
>  I guess there can't be an error that happens during supervisor termination that is not caused by one of its children.

That's on the spot! if we see a termination error that is not child related, then it is a serious bug